### PR TITLE
Rework ArgumentCodersCocoa to get us much closer to generated serializers

### PR DESCRIPTION
**Pull request by beidson**
#### f10ee49ef636c10dd181239c2b624bf85a954ebc
<pre>
Rework ArgumentCodersCocoa to get us much closer to generated serializers
<a href="https://bugs.webkit.org/show_bug.cgi?id=264588">https://bugs.webkit.org/show_bug.cgi?id=264588</a>
<a href="https://rdar.apple.com/118236121">rdar://118236121</a>

Reviewed by Alex Christensen.

This changes introduces ObjC object wrappers.
Encoding an ObjC object directly creates a typed wrapper around it.

For wrappers around a type that we still do directly via ObjC (e.g. NSArray), encoding the wrapper
then ends up going down the old &quot;encode ObjC directly&quot; path.

This change has the following benefits:
1 - We now broadly use serialization.in files for all ObjC types.
2 - For catch-all types like SecureCoding, we can now easily break out individual types into
    their own individual wrappers with individual serialization.
3 - So it follows, over time, the complicated logic in ArgumentCodersCocoa.mm can be whittled away.
4 - Having wrappers enables us to directly serialize collection types (like NSArray) without
    relying on the existing ObjC code path

* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.h:
* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.mm:
(WebCore::SerializedPlatformDataCueMac::allowedClassesForNativeValues):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeWithAllowedClasses):
(IPC::Decoder::allowedClasses const):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr):
(IPC::isObjectClassAllowed):
(IPC::decodeRequiringAllowedClasses):
(IPC::ArgumentCoder&lt;T::encode):
(IPC::ArgumentCoder&lt;CoreIPCRetainPtr&lt;T&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;CoreIPCRetainPtr&lt;T&gt;&gt;::decode):
(IPC::decodeWithAllowedClasses): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
(IPC::encodeObjectDirectly&lt;NSArray&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSArray&gt;):
(IPC::encodeObjectDirectly&lt;NSDictionary&gt;):
(IPC::id&gt;&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSDictionary&gt;):
(IPC::encodeObjectDirectly&lt;WebCore::CocoaFont&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;WebCore::CocoaFont&gt;):
(IPC::encodeObjectDirectly&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
(IPC::encodeObjectDirectly&lt;CFTypeRef&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;CFTypeRef&gt;):
(IPC::encodeObjectWithWrapper):
(IPC::decodeObjectFromWrapper):
(IPC::encodeArrayInternal): Deleted.
(IPC::decodeArrayInternal): Deleted.
(IPC::encodeColorInternal): Deleted.
(IPC::decodeColorInternal): Deleted.
(IPC::encodeDataInternal): Deleted.
(IPC::decodeDataInternal): Deleted.
(IPC::encodeDateInternal): Deleted.
(IPC::decodeDateInternal): Deleted.
(IPC::encodeDictionaryInternal): Deleted.
(IPC::decodeDictionaryInternal): Deleted.
(IPC::encodeFontInternal): Deleted.
(IPC::decodeFontInternal): Deleted.
(IPC::encodeNumberInternal): Deleted.
(IPC::decodeNumberInternal): Deleted.
(IPC::encodeSecureCodingInternal): Deleted.
(IPC::decodeSecureCodingInternal): Deleted.
(IPC::encodeStringInternal): Deleted.
(IPC::decodeStringInternal): Deleted.
(IPC::encodeURLInternal): Deleted.
(IPC::decodeURLInternal): Deleted.
(IPC::encodeCFInternal): Deleted.
(IPC::decodeCFInternal): Deleted.
(IPC::encodeObject): Deleted.
(IPC::decodeObject): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCCFType::CoreIPCCFType):
(WebKit::CoreIPCCFType::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCColor.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCColor::CoreIPCColor):
(WebKit::CoreIPCColor::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCColor.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCData.h:
(WebKit::CoreIPCData::CoreIPCData):
(WebKit::CoreIPCData::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDate.h:
(WebKit::CoreIPCDate::CoreIPCDate):
(WebKit::CoreIPCDate::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCFont.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCFont::CoreIPCFont):
(WebKit::CoreIPCFont::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCFont.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCData.h.
(WebKit::CoreIPCNSCFObject::CoreIPCNSCFObject):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm: Added.
(WebKit::valueFromID):
(WebKit::CoreIPCNSCFObject::CoreIPCNSCFObject):
(WebKit::CoreIPCNSCFObject::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCSecureCoding::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCString.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCString::CoreIPCString):
(WebKit::CoreIPCString::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCString.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCURL.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCURL::CoreIPCURL):
(WebKit::CoreIPCURL::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCURL.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/WebIconUtilities.mm:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
(WebKit::CoreIPCNumber::CoreIPCNumber):
(WebKit::CoreIPCNumber::toID):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder&lt;WebCore::SerializedPlatformDataCueValue&gt;::encodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::SerializedPlatformDataCueValue&gt;::decodePlatformData):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/270587@main">https://commits.webkit.org/270587@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0fe74be7745d6a7610495fa636417dfae964deb0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27948 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23667 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1885 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26100 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28528 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/26013 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23224 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29289 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23586 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23604 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27163 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2996 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1217 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4387 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6213 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3312 "Built successfully") | | |

### CODE REVIEW

**Comment by beidson** [comment]
This is to start an initial EWS pass, hence the lacking commit message.

File: Source/WebKit/Platform/IPC/Decoder.h [hunk 1]
```diff
@@ -185,6 +202,8 @@ class Decoder {
     ImportanceAssertion m_importanceAssertion;
 #endif
 
+    Vector<ClassStructPtr> m_allowedClasses;
```
Comment (achristensen07) [inline Source/WebKit/Platform/IPC/Decoder.h:207]: This could be inside PLATFORM(COCOA) since no other platforms decode ObjC types.

File: Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h [hunk 2]
```diff
@@ -32,16 +34,56 @@
...
+    {
+    }
+
+    CoreIPCRetainPtr(RetainPtr<T>&& object)
+        : RetainPtr<T>(WTFMove(object))
+    {
+    }
+};
+
+enum class NSType {
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h]: : uint8_t

File: Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+{
+}
+
+template<class... Ts>
+struct overload : Ts... {
+    using Ts::operator()...;
+};
+
+template<class... Ts>
+overload(Ts...) -> overload<Ts...>;
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm]: makeVisitor?  switchOn?

Comment (beidson) [reply Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm]: I...  Didn't know about those. Obviously way better.

File: Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm [hunk 4]
```diff
@@ -719,10 +648,12 @@ static bool shouldEnableStrictMode(Decoder& decoder, NSArray<Class> *allowedClas
     auto delegate = adoptNS([[WKSecureCodingArchivingDelegate alloc] init]);
     unarchiver.get().delegate = delegate.get();
 
-    auto allowedClassSet = adoptNS([[NSMutableSet alloc] initWithArray:allowedClasses]);
+    auto allowedClassSet = adoptNS([[NSMutableSet alloc] init]);
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm]: initWithCapacity would be slightly better here.

File: Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm [hunk 5]
```diff
@@ -719,10 +648,12 @@ static bool shouldEnableStrictMode(Decoder& decoder, NSArray<Class> *allowedClas
...
+    for (auto allowedClass : allowedClasses)
+        [allowedClassSet addObject:(allowedClass)];
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm]: parentheses not needed.

File: Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm [hunk 6]
```diff
@@ -669,45 +596,47 @@ static bool shouldEnableStrictMode(Decoder& decoder, NSArray<Class> *allowedClas
...
 #endif // ENABLE(APPLE_PAY)
-        || ([allowedClasses containsObject:NSURLCredential.class] && isDecodingKnownNSURLCredentialMessage(decoder))) // rdar://107553367
+        || (allowedClasses.contains(NSURLCredential.class) && isDecodingKnownNSURLCredentialMessage(decoder))) // rdar://107553367
         return true;
 
     // Note: Do not add more classes to the list of strict decoded classes.
     // If you want to serialize something new, extract its contents into a
     // struct and use a *.serialization.in file to serialize its contents.
-    RELEASE_LOG_FAULT(SecureCoding, "Strict mode check found unknown classes %@", allowedClasses);
+    RetainPtr<NSMutableArray> nsAllowedClasses = adoptNS([[NSMutableArray alloc] init]);
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm]: nit: initWithCapacity

File: Source/WebKit/Shared/Cocoa/CoreIPCCFType.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        : m_cfType(cfType)
+    {
+    }
+
+    CoreIPCCFType(RetainPtr<CFTypeRef>&& cfType)
+        : m_cfType(WTFMove(cfType))
+    {
+    }
+
+    RetainPtr<id> toID() { return static_cast<id>(m_cfType.get()); }
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCCFType.h]: (__bridge id) or bridge_cast?

File: Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#if PLATFORM(COCOA)
+
+webkit_platform_headers: "CoreIPCDictionary.h"
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:25]: I bet we could just change the python to add to the correct set of headers based on the presence of the [WebKitPlatform] attribute rather than include all these explicitly.

File: Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        CoreIPCColor,
+        CoreIPCData,
+        CoreIPCDate,
+        CoreIPCDictionary,
+        CoreIPCFont,
+        CoreIPCNumber,
+        CoreIPCSecureCoding,
+        CoreIPCString,
+        CoreIPCURL,
+        std::nullptr_t
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h]: This should probably be first in case the std::variant default constructor is called anywhere.

File: Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+RetainPtr<id> CoreIPCNSCFObject::toID()
+{
+    RetainPtr<id> result;
+
+    WTF::switchOn(m_value, [&](auto& object) {
+        result = object.toID();
+    }, [](std::nullptr_t) {
+        // result should be nil, which is the default value initialized above.
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:81]: ASSERT_NOT_REACHED()?

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
Comment (achristensen07) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Undo this change.

**Review by achristensen07 — APPROVED** [review]

File: Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm (hunk 10 above)
Comment (beidson) [reply Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:81]: The existing serialization code can actually encode/decode nil successfully, which this handles.

`void encodeObject(Encoder& encoder, id object)
 {
    encoder << static_cast<bool>(!object);
    if (!object)
        return;`
        
 And:
 
 `std::optional<RetainPtr<id>> decodeObject(Decoder& decoder, NSArray<Class> *allowedClasses)
 {
    bool isNull;
    if (!decoder.decode(isNull))
        return std::nullopt;
    if (isNull)
        return { nullptr };`

File: Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in (hunk 8 above)
Comment (beidson) [reply Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:25]: Yah, sounds reasonable.
Will do a targeted followup.

File: Source/WebKit/Shared/Cocoa/CoreIPCCFType.h (hunk 7 above)
Comment (beidson) [reply Source/WebKit/Shared/Cocoa/CoreIPCCFType.h]: I was going for identical code to `decodeCFInternal` here.

But agreed that's an improvement.
